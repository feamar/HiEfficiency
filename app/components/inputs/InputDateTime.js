import React, {Component} from  "react";
import {View} from "react-native";
import {Text} from "react-native-paper";
import PropTypes from "prop-types";
import {TouchableRipple} from "react-native-paper";
import DateTimePicker from 'react-native-modal-datetime-picker';
import UtilityTime from "../../utilities/UtilityTime";
import {asDate} from "../util/DateUtil";

const styles = {
    touchable:
    {
        flex: 1,
        paddingTop: 5,
        paddingBottom: 5,
        alignItems: "center",
        borderBottomWidth: 1,
        borderBottomColor: "#CECECE"
    }
}

export default class InputDateTime extends Component
{
    constructor(props)
    {
        super(props);

        this.state = {
            timestamp: this.props.timestamp,
            disabled: this.props.disabled,
            open: false,
            mode: this.props.mode
        }
    }

    onCancel = () =>
    {   this.close();}

    onConfirm = (timestamp) =>
    {
        this.close();
        this.setState({timestamp: timestamp}, () => 
        {
           if(this.props.onSelected)
           {    this.props.onSelected(timestamp);} 
        });
    }

    open = () =>
    {   this.setState({open: true});}
    
    close = () =>
    {   this.setState({open: false});}

    getDisplayValue = (timestamp, mode) =>
    {
        const date = new Date(timestamp);
        switch(mode)
        {
            case "date":
                return asDate(date);
            case "time":
                return UtilityTime.dateToHHMM(date);

            case "datetime":
                return this.getDisplayValue(timestamp, "date") + " " + this.getDisplayValue(timestamp, "time");
        }

        return "Undefined";
    }

    componentDidCatch(error, info)
    {
        this.setState({error: error});
    }

    render() 
    {
        return(
            <View testID="root" style={this.props.style}>
                <TouchableRipple testID="touchable" disabled={this.state.disabled} style={styles.touchable} onPress={this.open}><Text testID="text">{this.getDisplayValue(this.state.timestamp, this.state.mode)}</Text></TouchableRipple>
                {this.state.error == undefined && <DateTimePicker testID="picker" datePickerModeAndroid="default" date={new Date(this.state.timestamp)} onCancel={this.onCancel} mode={this.state.mode} isVisible={this.state.open} onConfirm={this.onConfirm} />}
            </View>
        );
    }
}

InputDateTime.defaultProps ={
    mode: "datetime"
}


InputDateTime.propTypes = {
    timestamp: PropTypes.any.isRequired,
    onSelected: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
    mode: PropTypes.string.isRequired
}