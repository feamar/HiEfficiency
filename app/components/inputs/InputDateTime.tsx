import React, {Component} from  "react";
import {View, StyleSheet, ViewStyle, TextStyle, ImageStyle, RegisteredStyle} from "react-native";
import {Text} from "react-native-paper";
import {TouchableRipple} from "react-native-paper";
import DateTimePicker from 'react-native-modal-datetime-picker';
import UtilityTime from "../../utilities/UtilityTime";
import {asDate} from "../util/DateUtil";

const styles = StyleSheet.create({
    touchable:
    {
        flex: 1,
        paddingTop: 5,
        paddingBottom: 5,
        alignItems: "center",
        width: "100%", 
        borderBottomWidth: 1,
        borderBottomColor: "#CECECE"
    }
});

type Mode = "date" | "time" | "datetime"  

interface Props
{
    timestamp: Date,
    disabled?: boolean,
    mode: Mode,
    onSelected: (timestamp: Date) => void,
    style?: RegisteredStyle<ViewStyle | TextStyle | ImageStyle>
}

interface State
{
    timestamp: Date,
    disabled: boolean,
    open: boolean,
    mode: Mode,
    error?: string
}


export default class InputDateTime extends Component<Props, State>
{
    constructor(props: Props)
    {
        super(props);

        this.state = {
            timestamp: this.props.timestamp,
            disabled: this.props.disabled || false,
            open: false,
            mode: this.props.mode
        }
    }

    onCancel = () =>
    {   this.close();}

    onConfirm = (timestamp: Date) =>
    {
        this.close();
        this.setState({timestamp: timestamp}, () => 
        {
           if(this.props.onSelected)
           {    this.props.onSelected(timestamp);} 
        });
    }

    open = () =>
    {   this.setState({open: true});}
    
    close = () =>
    {   this.setState({open: false});}

    getDisplayValue = (timestamp: Date, mode: Mode): string =>
    {
        const date = new Date(timestamp);
        switch(mode)
        {
            case "date":
                return asDate(date);
            case "time":
                return UtilityTime.dateToHHMM(date);

            case "datetime":
                return this.getDisplayValue(timestamp, "date") + " " + this.getDisplayValue(timestamp, "time");
        }

        return "Undefined";
    }

    componentDidCatch(error: Error, _: any)
    {   this.setState({error: error.toString()});}

    render() 
    {
        return(
            <View testID="root" style={this.props.style}>
                <TouchableRipple disabled={this.state.disabled} style={styles.touchable} onPress={this.open}><Text testID="text">{this.getDisplayValue(this.state.timestamp, this.state.mode)}</Text></TouchableRipple>
                {this.state.error == undefined && <DateTimePicker datePickerModeAndroid="spinner" date={new Date(this.state.timestamp)} onCancel={this.onCancel} mode={this.state.mode} isVisible={this.state.open} onConfirm={this.onConfirm} />}
            </View>
        );
    }
}
