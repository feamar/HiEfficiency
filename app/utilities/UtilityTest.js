

export default class UtilityText
{
    static findById = function(tree, testID) 
    {
        if(tree.props && tree.props.testID === testID) 
        {   return tree;}

        if(tree.children && tree.children.length > 0)
        {
            let childs = tree.children
            for(let i = 0; i < childs.length; i++)
            {
                let item = UtilityText.findById(childs[i], testID)
                if(typeof(item) !== 'undefined') 
                {   return item}
            }
        }
    }
}