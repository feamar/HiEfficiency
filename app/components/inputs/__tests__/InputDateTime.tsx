import React from "react";
import renderer from "react-test-renderer";
import InputDateTime from "../InputDateTime";
import UtilityTest from "../../../utilities/UtilityTest";


beforeAll(() => 
{
    jest.mock('Platform', () => {
        const Platform = require('Platform');
        Platform.OS = 'android';
        return Platform;
    });
});

describe("An InputDateTime", () => 
{
    const timestamp = new Date(1514761200000); //2018-01-01
    const onSelectedMock = jest.fn();
  
    let component = renderer.create(<InputDateTime mode="date" timestamp={timestamp} onSelected={onSelectedMock} />);
    let tree = component.toJSON();
    let instance = component.getInstance();

    test("Opening the picker" , () => 
    {
        instance.open();
        expect(instance.state.open).toBe(true);

        instance.close();
        expect(instance.state.open).toBe(false);
    }); 

    test("Mode defaults to 'datetime'", () => 
    {   expect(instance.state.mode).toBe("datetime");});

    test("Elements should be present", () => 
    {
        expect(UtilityTest.findById(tree, "root")).toBeDefined();
        expect(UtilityTest.findById(tree, "touchable")).toBeDefined();
        expect(UtilityTest.findById(tree, "text")).toBeDefined();
    });

    test("Display value is correct", () => 
    {
        expect(instance.getDisplayValue(timestamp, "datetime")).toBe("Jan 01 2018 00:00");
        expect(instance.getDisplayValue(timestamp, "time")).toBe("00:00");
        expect(instance.getDisplayValue(timestamp, "date")).toBe("Jan 01 2018");
    });

    test("Callback onSelected is called", () => 
    {
        instance.onConfirm(1529857953000) //2018-06-24 at 18:32:33
        expect(onSelectedMock).toBeCalledWith(1529857953000);
        expect(onSelectedMock).toHaveBeenCalledTimes(1);
    });

    test("Renders correctly", () => 
    {
        expect(tree).toMatchSnapshot();
    });

});

