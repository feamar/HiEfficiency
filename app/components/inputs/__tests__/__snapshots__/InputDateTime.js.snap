// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An InputDateTime Renders correctly 1`] = `
<View
  testID="root"
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#CECECE",
        "borderBottomWidth": 1,
        "flex": 1,
        "paddingBottom": 5,
        "paddingTop": 5,
      }
    }
    testID="touchable"
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          null,
        ]
      }
      testID="text"
      theme={
        Object {
          "colors": Object {
            "accent": "#03dac4",
            "backdrop": "rgba(0, 0, 0, 0.5)",
            "background": "#f6f6f6",
            "disabled": "rgba(0, 0, 0, 0.26)",
            "error": "#B00020",
            "placeholder": "rgba(0, 0, 0, 0.54)",
            "primary": "#6200ee",
            "surface": "#ffffff",
            "text": "#000000",
          },
          "dark": false,
          "fonts": Object {
            "light": "HelveticaNeue-Light",
            "medium": "HelveticaNeue-Medium",
            "regular": "Helvetica Neue",
            "thin": "HelveticaNeue-Thin",
          },
          "roundness": 4,
        }
      }
    >
      Jan 01 2018 00:00
    </Text>
  </View>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onRequestClose={[Function]}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      pointerEvents="box-none"
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-end",
          "margin": 10,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "white",
              "borderRadius": 13,
              "marginBottom": 8,
              "overflow": "hidden",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderBottomColor": "#d5d5d5",
              "borderBottomWidth": 0.5,
              "padding": 14,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#8f8f8f",
                  "fontSize": 13,
                  "textAlign": "center",
                },
                undefined,
              ]
            }
          >
            Pick a date
          </Text>
        </View>
        <View
          onStartShouldSetResponderCapture={[Function]}
        >
          <View>
            <RCTDatePicker
              date={1514761200000}
              minuteInterval={1}
              mode="datetime"
              onChange={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "height": 216,
                }
              }
            />
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "#d5d5d5",
              "borderTopWidth": 0.5,
              "height": 57,
              "justifyContent": "center",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "#007ff9",
                  "fontSize": 20,
                  "fontWeight": "normal",
                  "textAlign": "center",
                },
                undefined,
              ]
            }
          >
            Confirm
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 13,
            "height": 57,
            "justifyContent": "center",
            "marginBottom": 0,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#007ff9",
                "fontSize": 20,
                "fontWeight": "600",
                "padding": 10,
                "textAlign": "center",
              },
              undefined,
            ]
          }
        >
          Cancel
        </Text>
      </View>
    </View>
  </Modal>
</View>
`;
